students/ThanhNhan/lesson09/mail5.py

```diff
@@ -0,0 +1,197 @@
+
+    def ck_dir(self):
+        if not os.path.isdir(self.DIR_NAME):
+            os.mkdir(self.DIR_NAME)
+            return False
+        return True
+
+    def ss_letterall(self):
+    #     """
+    #     This method write letter to disk as txt file
+
+    #     params:  none
+    #     will check and generate directory
+    #     will write all txt file in the directory
+    #     """
+
+        #writes each letter to disk as txt file
+        #open ("./thankyou.txt", "w")
+        # tempfile.gettempdir() 
+
+        for key, values in self.data_collection.items():  
+            d = Donor(key)
+            letter = d.generate_letter(values)
+            filename = key + ".txt"
+            self.ck_dir()    
+            try:
+                print(filename) 
+                filename = os.path.join(self.DIR_NAME, filename)
+                open(filename, 'w').write(letter)
+            except FileNotFoundError:
+                print("couldn't open missing.txt")
+            finally:
+                pass
+    
+
+    def add_donor(self, Donor):
+        if self.find_donor() is True:
+            self.data_collection[Donor.name].append(Donor.donations[0])    
+        else:
+            self.data_collection[Donor.name] = Donor.donations
+
+
+
+class Donor(DonorCollection):
+
+    def __init__(self, name, donations=None):  
+        """
+        Create Donor object
+        param name: the full name of the donor
+        param donations: the donations 
+        """
+
+        self.name = name
+        if donations is None:
+            self.donations = []
+        else:
+            self.donations = [donations]
+        print("name is {}".format(self.name))
+        print("donations is", self.donations)
+
+    # @property #outside look like attribute
+    def add_donations(self, amount):
+        """
+        Add donation amount of user input
+        """
+        donate_amount = self._ck_dollar(amount)
+        self.donations.append(donate_amount)
+    
+    @property #outside look like attribute
+    def num_donations(self):
+        """
+        measure 
+        param donations: iterable donations
+        """
+        return len(self.donations)
+
+
+    @staticmethod
+    def _ck_dollar(d_amount):
+        """
+        This method is check if user input correct float number.
+        Check if user not a nubmber or infinity number
+        Check if user enter negative amount
+
+        params amount:  input amount of dollar value
+        returns amount dollar value
+        """
+
+        amount_flt = round(float(d_amount), 2)
+
+        if math.isnan(amount_flt) or math.isinf(amount_flt):
+            raise ValueError("Is not a number or infinite number")
+
+        if amount_flt <= 0.00:
+            raise ValueError("Can't be negative number")
+            
+        return amount_flt
+
+    
+    def validate_name(self):
+        """
+        This method is validate user name input
+
+        params:  input user name
```

Comment:
again: no parameters -- it is acting on self -- none to pass in.

though it would probably be better if you DID pass in the name to be validated -- so you could validate before setting it on the instance. LIke `_ck_dollar`